### PR TITLE
Adds `LazyRawTextReader` support for reading lists

### PR DESCRIPTION
**Pull request by zslayton**
Builds on outstanding PRs #609, #612, #613, #614 and #616.

Adds support for reading lists to the `LazyRawTextReader`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lazy/binary/raw/sequence.rs [hunk 1]
```diff
@@ -18,11 +18,11 @@ impl<'data> LazyRawBinarySequence<'data> {
         self.value.ion_type()
     }
 
-    pub fn iter(&self) -> RawSequenceIterator<'data> {
+    pub fn iter(&self) -> RawBinarySequenceIterator<'data> {
         // Get as much of the sequence's body as is available in the input buffer.
         // Reading a child value may fail as `Incomplete`
         let buffer_slice = self.value.available_body();
-        RawSequenceIterator::new(buffer_slice)
+        RawBinarySequenceIterator::new(buffer_slice)
```
Comment (zslayton) [inline src/lazy/binary/raw/sequence.rs:25]: 🗺️ Now that there are both binary and text readers, `RawSequenceIterator` has become the `RawBinarySequenceIterator`.

File: src/lazy/text/buffer.rs
Comment (zslayton) [inline src/lazy/text/buffer.rs:1]: 🗺️ This file contains the new parsing logic needed to detect the beginning of a list, a value in the list, and the end of the list.

File: src/lazy/text/matched.rs
Comment (zslayton) [inline src/lazy/text/matched.rs:1]: 🗺️ All changes in this file stem from the change on line 175; see the `TODO` comment.

File: src/lazy/text/raw/sequence.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.input.match_list_value() {
+            Ok((remaining, Some(value))) => {
+                self.input = remaining;
+                Some(Ok(value))
+            }
+            Ok((_remaining, None)) => None,
+            Err(e) => e
+                .with_context("reading the next list value", self.input)
+                .transpose(),
```
Comment (zslayton) [inline src/lazy/text/raw/sequence.rs:123]: 🗺️ This call to `transpose()` is the `ToIteratorOutput` trait's extension method.

**Review by zslayton — COMMENTED** [review]
🗺️ PR Tour

File: src/lazy/text/raw/sequence.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+impl<'a, 'data> IntoIterator for &'a LazyRawTextSequence<'data> {
+    type Item = IonResult<LazyRawTextValue<'data>>;
+    type IntoIter = RawTextSequenceIterator<'data>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+impl<'a> Debug for LazyRawTextSequence<'a> {
```
[the remark this answers is not in the document]
Comment (popematt) [reply src/lazy/text/raw/sequence.rs:63]: Suggestion—can we just use `self.value.input.data` (assuming it's valid utf-8) and if it fails, then we fallback to the `Debug` impl of the `TextBufferView`? This seems like an approach that could work for _all_ lazy text values. What do you think?

**Review by popematt — APPROVED** [review]

**Pull request merged**